neqo-http3/src/control_stream_remote.rs

```diff
@@ -24,45 +25,51 @@ impl ::std::fmt::Display for ControlStreamRemote {
 }
 
 impl ControlStreamRemote {
-    pub fn new() -> Self {
+    pub fn new(stream_id: u64) -> Self {
         Self {
-            stream_id: None,
+            stream_id,
             frame_reader: HFrameReader::new(),
-            fin: false,
         }
     }
 
-    /// A remote control stream has been received. Inform `ControlStreamRemote`.
-    pub fn add_remote_stream(&mut self, stream_id: u64) -> Res<()> {
-        qinfo!([self], "A new control stream {}.", stream_id);
-        if self.stream_id.is_some() {
-            qdebug!([self], "A control stream already exists");
-            return Err(Error::HttpStreamCreation);
+    /// Check if a stream is the control stream and read received data.
+    pub fn receive_single(&mut self, conn: &mut Connection) -> Res<Option<HFrame>> {
+        qdebug!([self], "Receiving data.");
+        match self.frame_reader.receive(conn, self.stream_id)? {
+            (_, true) => Err(Error::HttpClosedCriticalStream),
+            (s, false) => Ok(s),
         }
-        self.stream_id = Some(stream_id);
-        Ok(())
     }
+}
 
-    /// Check if `stream_id` is the remote control stream.
-    pub fn is_recv_stream(&self, stream_id: u64) -> bool {
-        matches!(self.stream_id, Some(id) if id == stream_id)
+impl RecvStream for ControlStreamRemote {
+    fn stream_reset(&self, _error: AppError, _reset_type: ResetType) -> Res<()> {
+        Err(Error::HttpClosedCriticalStream)
     }
 
-    /// Check if a stream is the control stream and read received data.
-    pub fn receive(&mut self, conn: &mut Connection) -> Res<Option<HFrame>> {
-        assert!(self.stream_id.is_some());
-        qdebug!([self], "Receiving data.");
-        match self.frame_reader.receive(conn, self.stream_id.unwrap())? {
-            (_, true) => {
-                self.fin = true;
-                Err(Error::HttpClosedCriticalStream)
+    fn receive(&mut self, conn: &mut Connection) -> Res<ReceiveOutput> {
+        let mut control_frames = Vec::new();
+
+        loop {
+            if let Some(f) = self.receive_single(conn)? {
+                control_frames.push(f);
+            } else if control_frames.is_empty() {
+                return Ok(ReceiveOutput::NoOutput);
```

Comment:
Do you need to worry about this?  If there are no control frames, that's OK, right?  The caller will just do nothing.

Reply:
not really. We can just return an empty vec.